File: metro_tracker_app/lib/features/metro_tracking/domain/entities/vehicle.dart
import 'package:equatable/equatable.dart';

enum VehicleType { bus, metro, unknown }

class Vehicle extends Equatable {
  final String id;
  final VehicleType type;
  final String routeId;
  final double latitude;
  final double longitude;
  final double bearing;
  final double speed;
  final DateTime lastUpdate;

  const Vehicle({
    required this.id,
    required this.type,
    required this.routeId,
    required this.latitude,
    required this.longitude,
    required this.bearing,
    required this.speed,
    required this.lastUpdate,
  });

  @override
  List<Object?> get props =>
      [id, type, routeId, latitude, longitude, bearing, speed, lastUpdate];
}

File: backend/src/models/Vehicle.js
const mongoose = require('mongoose');

const vehicleSchema = new mongoose.Schema({
  vehicleId: {
    type: String,
    required: true,
    unique: true,
    trim: true
  },
  type: {
    type: String,
    enum: ['METRO', 'BUS'],
    required: true
  },
  route: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Route',
    required: true
  },
  currentStation: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Station',
    required: true
  },
  nextStation: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Station'
  },
  // Progress between current and next station (0.0 to 1.0)
  progressToNextStation: {
    type: Number,
    default: 0,
    min: 0,
    max: 1
  },
  active: {
    type: Boolean,
    default: true
  },
  lastUpdate: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Index for faster queries

vehicleSchema.index({ route: 1, active: 1 });
vehicleSchema.index({ active: 1 });

// Method to convert to JSON for WebSocket transmission
vehicleSchema.methods.toClientJSON = function () {
  return {
    id: this.vehicleId,
    type: this.type,
    routeId: this.route,
    currentStationId: this.currentStation,
    nextStationId: this.nextStation,
    progress: this.progressToNextStation,
    timestamp: Math.floor(this.lastUpdate.getTime() / 1000)
  };
};

module.exports = mongoose.model('Vehicle', vehicleSchema);
